### ISSUE TITLE
New driver support (version 510.47.03)

### ISSUE THREAD

**Issue by LionHeartP** [keylase/nvidia-patch#493]
**Please provide details about your environment (please complete the following information)**
- Nvidia driver version: 510.47.03
- GPU Model: RTX 3070
- OS: Arch Linux
- [x] I'm willing to participate in testing of new patch version

**Comment by LionHeartP** [comment]
Just as an FYI I tried using the previous addresses to patch this version but it didn't work, at least for the FBC one.

**Comment by mauricioquintela** [comment]
I tried it with the addresses listed in https://github.com/keylase/nvidia-patch/issues/487 (updating of course ["510.39.01"] to ["510.47.03"]), and fbc is working fine for me. Haven't tested nvenc yet.

**Comment by nathan-t-sudo** [comment]
added the following patch on my end
and its working fine

diff --git a/drivers.json b/drivers.json
index 1e5990b..f801a10 100644
--- a/drivers.json
+++ b/drivers.json
@@ -667,6 +667,12 @@
```

                     "nvenc_patch": true,
                     "nvfbc_patch": true,
                     "driver_url": "https://international.download.nvidia.com/XFree86/Linux-x86_64/470.94/NVIDIA-Linux-x86_64-470.94.run"

+                },
+                {
+                    "version": "510.47.03",
+                    "nvenc_patch": true,
+                    "nvfbc_patch": true,
+          
          "driver_url": "https://download.nvidia.com/XFree86/Linux-x86_64/510.47.03/NVIDIA-Linux-x86_64-510.47.03.run"
                 }

             ],
             "example": {
              
```
diff --git a/patch.sh b/patch.sh
index 690b009..9dd4cd5 100755
--- a/patch.sh
+++ b/patch.sh
@@ -173,6 +173,7 @@ declare -A patch_list=(
     ["495.29.05"]='s/\xe8\x35\x1f\xff\xff\x85\xc0\x41\x89\xc4/\xe8\x35\x1f\xff\xff\x29\xc0\x41\x89\xc4/g'
     ["495.44"]='s/\xe8\x35\x1f\xff\xff\x85\xc0\x41\x89\xc4/\xe8\x35\x1f\xff\xff\x29\xc0\x41\x89\xc4/g'
     ["495.46"]='s/\xe8\x35\x1f\xff\xff\x85\xc0\x41\x89\xc4/\xe8\x35\x1f\xff\xff\x29\xc0\x41\x89\xc4/g'
`+    ["510.47.03"]='s/\xe8\x35\x1f\xff\xff\x85\xc0\x41\x89\xc4/\xe8\x35\x1f\xff\xff\x29\xc0\x41\x89\xc4/g'`
 )
 
 declare -A object_list=(
@@ -292,6 +293,8 @@ declare -A object_list=(
     ["495.29.05"]='libnvidia-encode.so'
     ["495.44"]='libnvidia-encode.so'
     ["495.46"]='libnvidia-encode.so'
`+    ["510.47.03"]='libnvidia-encode.so'`

 )
 
 check_version_supported () {

diff --git a/patch-fbc.sh b/patch-fbc.sh
index 2eced4c..41bc841 100755
--- a/patch-fbc.sh
+++ b/patch-fbc.sh
@@ -135,6 +135,7 @@ declare -A patch_list=(
     ["495.29.05"]='s/\x83\xfe\x01\x73\x08\x48/\x83\xfe\x00\x72\x08\x48/'
     ["495.44"]='s/\x83\xfe\x01\x73\x08\x48/\x83\xfe\x00\x72\x08\x48/'
     ["495.46"]='s/\x83\xfe\x01\x73\x08\x48/\x83\xfe\x00\x72\x08\x48/'
`+    ["510.47.03"]='s/\x83\xfe\x01\x73\x08\x48/\x83\xfe\x00\x72\x08\x48/'`
 )
 
 declare -A object_list=(
@@ -218,6 +219,7 @@ declare -A object_list=(
     ["495.29.05"]='libnvidia-fbc.so'
     ["495.44"]='libnvidia-fbc.so'
     ["495.46"]='libnvidia-fbc.so'
`+    ["510.47.03"]='libnvidia-fbc.so'`
 )
 
 check_version_supported () {

**Comment by LionHeartP** [comment]
> added the following patch on my end and its working fine
> 
> diff --git a/drivers.json b/drivers.json index 1e5990b..f801a10 100644 --- a/drivers.json +++ b/drivers.json @@ -667,6 +667,12 @@
> 
> ```
> 
>                      "nvenc_patch": true,
>                      "nvfbc_patch": true,
>                      "driver_url": "https://international.download.nvidia.com/XFree86/Linux-x86_64/470.94/NVIDIA-Linux-x86_64-470.94.run"
> 
> +                },
> +                {
> +                    "version": "510.47.03",
> +                    "nvenc_patch": true,
> +                    "nvfbc_patch": true,
> +          
>           "driver_url": "https://download.nvidia.com/XFree86/Linux-x86_64/510.47.03/NVIDIA-Linux-x86_64-510.47.03.run"
>                  }
> 
>              ],
>              "example": {
>               
> ```
> 
> diff --git a/patch.sh b/patch.sh index 690b009..9dd4cd5 100755 --- a/patch.sh +++ b/patch.sh @@ -173,6 +173,7 @@ declare -A patch_list=( ["495.29.05"]='s/\xe8\x35\x1f\xff\xff\x85\xc0\x41\x89\xc4/\xe8\x35\x1f\xff\xff\x29\xc0\x41\x89\xc4/g' ["495.44"]='s/\xe8\x35\x1f\xff\xff\x85\xc0\x41\x89\xc4/\xe8\x35\x1f\xff\xff\x29\xc0\x41\x89\xc4/g' ["495.46"]='s/\xe8\x35\x1f\xff\xff\x85\xc0\x41\x89\xc4/\xe8\x35\x1f\xff\xff\x29\xc0\x41\x89\xc4/g' `+ ["510.47.03"]='s/\xe8\x35\x1f\xff\xff\x85\xc0\x41\x89\xc4/\xe8\x35\x1f\xff\xff\x29\xc0\x41\x89\xc4/g'` )
> 
> declare -A object_list=( @@ -292,6 +293,8 @@ declare -A object_list=( ["495.29.05"]='libnvidia-encode.so' ["495.44"]='libnvidia-encode.so' ["495.46"]='libnvidia-encode.so' `+ ["510.47.03"]='libnvidia-encode.so'`
> 
> )
> 
> check_version_supported () {
> 
> diff --git a/patch-fbc.sh b/patch-fbc.sh index 2eced4c..41bc841 100755 --- a/patch-fbc.sh +++ b/patch-fbc.sh @@ -135,6 +135,7 @@ declare -A patch_list=( ["495.29.05"]='s/\x83\xfe\x01\x73\x08\x48/\x83\xfe\x00\x72\x08\x48/' ["495.44"]='s/\x83\xfe\x01\x73\x08\x48/\x83\xfe\x00\x72\x08\x48/' ["495.46"]='s/\x83\xfe\x01\x73\x08\x48/\x83\xfe\x00\x72\x08\x48/' `+ ["510.47.03"]='s/\x83\xfe\x01\x73\x08\x48/\x83\xfe\x00\x72\x08\x48/'` )
> 
> declare -A object_list=( @@ -218,6 +219,7 @@ declare -A object_list=( ["495.29.05"]='libnvidia-fbc.so' ["495.44"]='libnvidia-fbc.so' ["495.46"]='libnvidia-fbc.so' `+ ["510.47.03"]='libnvidia-fbc.so'` )
> 
> check_version_supported () {

Confirmed working fine.

**Comment by LionHeartP** [comment]
Created a PR with this patch. Please test. 

https://github.com/keylase/nvidia-patch/pull/494

**Comment by didyouexpectthat** [comment]
This doesn't work for me.

```
# ./patch.sh
Detected nvidia driver version: 510.47.03
Attention! Backup not found. Copying current libnvidia-encode.so to backup.
09ff74cd645a2309afebf21a28f5a42934e0a0f7  /opt/nvidia/libnvidia-encode-backup/libnvidia-encode.so.510.47.03
09ff74cd645a2309afebf21a28f5a42934e0a0f7  /usr/lib/x86_64-linux-gnu/libnvidia-encode.so.510.47.03
Patched!
```
Old and new are the same. 

```
# ./patch-fbc.sh
Detected nvidia driver version: 510.47.03
Attention! Backup not found. Copying current libnvidia-fbc.so to backup.
e58b9b7e6f4e55fc7c6e077ef534c5bd70576d92  /opt/nvidia/libnvidia-fbc-backup/libnvidia-fbc.so.510.47.03
4c8571afd264a74321d9f2ef6409f70024f2858b  /usr/lib/x86_64-linux-gnu/libnvidia-fbc.so.510.47.03
Patched!
```
This seems to have done something.

**Comment by dor7to** [comment]
> Created a PR with this patch. Please test.
> 
> #494

works good on my end, cheers.

```
# ./patch.sh 
Detected nvidia driver version: 510.47.03
Attention! Backup not found. Copying current libnvidia-encode.so to backup.
09ff74cd645a2309afebf21a28f5a42934e0a0f7  /opt/nvidia/libnvidia-encode-backup/libnvidia-encode.so.510.47.03
09ff74cd645a2309afebf21a28f5a42934e0a0f7  /usr/lib64/libnvidia-encode.so.510.47.03
Patched!
```

```
# ./patch-fbc.sh 
Detected nvidia driver version: 510.47.03
Attention! Backup not found. Copying current libnvidia-fbc.so to backup.
e58b9b7e6f4e55fc7c6e077ef534c5bd70576d92  /opt/nvidia/libnvidia-fbc-backup/libnvidia-fbc.so.510.47.03
4c8571afd264a74321d9f2ef6409f70024f2858b  /usr/lib64/libnvidia-fbc.so.510.47.03
Patched!
```

**Comment by didyouexpectthat** [comment]
> > Created a PR with this patch. Please test.
> > #494
> 
> works good on my end, cheers.
> 
> ```
> # ./patch.sh 
> Detected nvidia driver version: 510.47.03
> Attention! Backup not found. Copying current libnvidia-encode.so to backup.
> 09ff74cd645a2309afebf21a28f5a42934e0a0f7  /opt/nvidia/libnvidia-encode-backup/libnvidia-encode.so.510.47.03
> 09ff74cd645a2309afebf21a28f5a42934e0a0f7  /usr/lib64/libnvidia-encode.so.510.47.03
> Patched!
> ```
> 
> ```
> # ./patch-fbc.sh 
> Detected nvidia driver version: 510.47.03
> Attention! Backup not found. Copying current libnvidia-fbc.so to backup.
> e58b9b7e6f4e55fc7c6e077ef534c5bd70576d92  /opt/nvidia/libnvidia-fbc-backup/libnvidia-fbc.so.510.47.03
> 4c8571afd264a74321d9f2ef6409f70024f2858b  /usr/lib64/libnvidia-fbc.so.510.47.03
> Patched!
> ```

Your checksum is the same also for the patch. I don't think it successfully patched, otherwise, the checksum would be different, like on patch-fbc.

**Comment by Czar102** [comment]
Same as above
https://github.com/keylase/nvidia-patch/pull/494#issuecomment-1033849984

**Comment by LionHeartP** [comment]
I would suggest to the people that can't get patch.sh working, to maybe try out this.
https://github.com/illnyang/nvlax
If you are on Arch, prepatched drivers already exist on AUR as well as nvlax itself. You can also make it work with flatpak apps by pointing it to the runtime .so files.

I think something has changed on Nvidia's end, and we need different addresses for patch.sh

**Comment by spider3000** [comment]
> If you are on Arch, prepatched drivers already exist on AUR

What's the package called?

**Comment by Czar102** [comment]
Thanks a lot for help!
I tried to use https://github.com/illnyang/nvlax, but I ran into some compilation errors.

This has already been observed: https://github.com/illnyang/nvlax/issues/11

There is a pull request made by [tytan652](https://github.com/tytan652) that aims to fix this error.
He also made an AUR package -- https://aur.archlinux.org/packages/nvidia-utils-nvlax
But I am still having compilation errors, even in the patched repos/AUR packages...

**Comment by LionHeartP** [comment]
> Thanks a lot for help! I tried to use https://github.com/illnyang/nvlax, but I ran into some compilation errors.
> 
> This has already been observed: [illnyang/nvlax#11](https://github.com/illnyang/nvlax/issues/11)
> 
> There is a pull request made by [tytan652](https://github.com/tytan652) that aims to fix this error. He also made an AUR package -- https://aur.archlinux.org/packages/nvidia-utils-nvlax But I am still having compilation errors, even in the patched repo/AUR package...

Tytan also made an nvlax package for AUR that also fixes the compilation error if you're interested. It's called nvlax-git

**Comment by ConstBur** [comment]
> > I get a compilation error with nvlax-git as well, here are the logs where it crashed:
> > ```
> > [ 95%] Building CXX object CMakeFiles/nvlax_encode.dir/src/lax_encode.cc.o
> > /var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/src/lax_fbc.cc: In function ‘int main(int, char**)’:
> > /var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/src/lax_fbc.cc:40:27: error: request for member ‘virtual_address’ in ‘s_rodata’, which is of pointer type ‘LIEF::ELF::Section*’ (maybe you meant to use ‘->’ ?)
> >    40 |         offset = s_rodata.virtual_address() + s_rodata.search("This hardware does not support NvFBC");
> >       |                           ^~~~~~~~~~~~~~~
> > /var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/src/lax_fbc.cc:40:56: error: request for member ‘search’ in ‘s_rodata’, which is of pointer type ‘LIEF::ELF::Section*’ (maybe you meant to use ‘->’ ?)
> >    40 |         offset = s_rodata.virtual_address() + s_rodata.search("This hardware does not support NvFBC");
> >       |                                                        ^~~~~~
> > /var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/src/lax_fbc.cc:52:38: error: request for member ‘content’ in ‘s_text’, which is of pointer type ‘LIEF::ELF::Section*’ (maybe you meant to use ‘->’ ?)
> >    52 |         auto v_text_content = s_text.content();
> >       |                                      ^~~~~~~
> > /var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/src/lax_fbc.cc:61:38: error: request for member ‘virtual_address’ in ‘s_text’, which is of pointer type ‘LIEF::ELF::Section*’ (maybe you meant to use ‘->’ ?)
> >    61 |                 size_t temp = s_text.virtual_address() +
> >       |                                      ^~~~~~~~~~~~~~~
> > /var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/src/lax_fbc.cc:67:37: error: request for member ‘virtual_address’ in ‘s_text’, which is of pointer type ‘LIEF::ELF::Section*’ (maybe you meant to use ‘->’ ?)
> >    67 |                     offset = s_text.virtual_address() + data - v_text_content.data();
> >       |                                     ^~~~~~~~~~~~~~~
> > /var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/src/lax_encode.cc: In function ‘void patch_linux(LIEF::ELF::Binary*)’:
> > /var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/src/lax_encode.cc:45:78: error: request for member ‘value’ in ‘f_nvenc_ci’, which is of pointer type ‘LIEF::Symbol*’ (maybe you meant to use ‘->’ ?)
> >    45 |         auto v_func_bytes = bin->get_content_from_virtual_address(f_nvenc_ci.value(), 0x260);
> >       |                                                                              ^~~~~
> > /var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/src/lax_encode.cc:53:37: error: request for member ‘value’ in ‘f_nvenc_ci’, which is of pointer type ‘LIEF::Symbol*’ (maybe you meant to use ‘->’ ?)
> >    53 |                 offset = f_nvenc_ci.value() +
> >       |                                     ^~~~~
> > make[2]: *** [CMakeFiles/nvlax_fbc.dir/build.make:90: CMakeFiles/nvlax_fbc.dir/src/lax_fbc.cc.o] Error 1
> > make[2]: Leaving directory '/var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/build'
> > make[1]: *** [CMakeFiles/Makefile2:219: CMakeFiles/nvlax_fbc.dir/all] Error 2
> > make[1]: *** Waiting for unfinished jobs....
> > make[2]: *** [CMakeFiles/nvlax_encode.dir/build.make:90: CMakeFiles/nvlax_encode.dir/src/lax_encode.cc.o] Error 1
> > make[2]: Leaving directory '/var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/build'
> > make[1]: *** [CMakeFiles/Makefile2:190: CMakeFiles/nvlax_encode.dir/all] Error 2
> > make[1]: Leaving directory '/var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/build'
> > make: *** [Makefile:136: all] Error 2
> > make: Leaving directory '/var/tmp/pamac-build-constbur/nvlax-git/src/nvlax/build'
> > ==> ERROR: A failure occurred in build().
> >     Aborting...
> > ```
> 
> Ah pamac. This must be Manjaro? I compiled it successfully like 3-4 days ago on arch using yay. Maybe Manjaro is missing something from the latest arch repos?

Nah, I installed it myself on Arco :) 
I've tried with yay, paru and compiling directly from source with the patch and still get the same thing so yeah...

**Comment by Czar102** [comment]
I got a slightly different error...

```
[ 91%] Building CXX object CMakeFiles/nvlax_encode.dir/src/lax_encode.cc.o
/nvlax-git/src/nvlax/src/lax_encode.cc: In function 'void patch_linux(LIEF::ELF::Binary*)':
/nvlax-git/src/nvlax/src/lax_encode.cc:45:78: error: request for member 'value' in 'f_nvenc_ci', which is of pointer type 'LIEF::Symbol*' (maybe you meant to use '->' ?)
   45 |         auto v_func_bytes = bin->get_content_from_virtual_address(f_nvenc_ci.value(), 0x260);
      |                                                                              ^~~~~
/nvlax-git/src/nvlax/src/lax_encode.cc:53:37: error: request for member 'value' in 'f_nvenc_ci', which is of pointer type 'LIEF::Symbol*' (maybe you meant to use '->' ?)
   53 |                 offset = f_nvenc_ci.value() +
      |                                     ^~~~~
make[2]: *** [CMakeFiles/nvlax_encode.dir/build.make:90: CMakeFiles/nvlax_encode.dir/src/lax_encode.cc.o] Error 1
make[2]: Leaving directory '/nvlax-git/src/nvlax/build'
make[1]: *** [CMakeFiles/Makefile2:190: CMakeFiles/nvlax_encode.dir/all] Error 2
make[1]: Leaving directory '/nvlax-git/src/nvlax/build'
make: *** [Makefile:136: all] Error 2
```

**Comment by leillo1975** [comment]
patch-fbc.sh works for me, but patch.sh give me the following error:
`Patch for this (510.47.03) nvidia driver not found.`

**Comment by LionHeartP** [comment]
> patch-fbc.sh works for me, but patch.sh give me the following error: `Patch for this (510.47.03) nvidia driver not found.`

Yes I removed nvenc patch because it uses different addresses. Fbc should work fine.

**Comment by Czar102** [comment]
https://aur.archlinux.org/packages/nvidia-utils-nvlax

The AUR package has been updated and worked for me!

**Comment by robkorv** [comment]
Ubuntu 20.04.3 LTS ✅  for the "patch-fbc.sh" and fbc works in obs-studio now. Tnx!

```shell
➜  nvidia-patch git:(LionHeartP/master) sudo ./patch-fbc.sh                                                                                                                                                                                                     
Detected nvidia driver version: 510.47.03
Attention! Backup not found. Copying current libnvidia-fbc.so to backup.
e58b9b7e6f4e55fc7c6e077ef534c5bd70576d92  /opt/nvidia/libnvidia-fbc-backup/libnvidia-fbc.so.510.47.03
4c8571afd264a74321d9f2ef6409f70024f2858b  /usr/lib/x86_64-linux-gnu/libnvidia-fbc.so.510.47.03
Patched!
```

**Comment by Kyhze** [comment]
> 

Any alternatives for debian users?

**Comment by jailuthra** [comment]
Hey all fixed in #499, please test :)
